stan/math/prim/fun/tcrossprod.hpp

```diff
@@ -11,21 +12,24 @@ namespace math {
  * Returns the result of post-multiplying a matrix by its
  * own transpose.
  *
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T type of the matrix (must be derived from \c Eigen::MatrixBase)
  * @param M Matrix to multiply.
  * @return M times its transpose.
  */
-template <int R, int C>
-inline Eigen::MatrixXd tcrossprod(const Eigen::Matrix<double, R, C>& M) {
+template <typename T, require_eigen_vt<std::is_arithmetic, T>* = nullptr>
+inline Eigen::Matrix<value_type_t<T>, T::RowsAtCompileTime,
+                     T::RowsAtCompileTime>
```

Comment:
[optional]

If you want to make this a bit prettier I think you could declare this type in the template like

```
, ReturnType = Eigen::Matrix<value_type_t<T>, T::RowsAtCompileTime,
                     T::RowsAtCompileTime>>
```

Reply:
Personally I would prefer not to put another thing in template lists. They already contain actual templates and requires. So in this case I think the code being clearer outweighs some code duplication. If you disagree I can still change that.